File: apps/python/lib/chapter10/classifier.py
from dataclasses import dataclass
from typing import Protocol, Self

import pandas as pd

from lib.chapter01.kinoko_takenoko import accuracy
from lib.chapter02.iris_preprocessing import TrainTestSplit


class Classifier(Protocol):
    def fit(self, x: pd.DataFrame, t: pd.Series) -> Self: ...

    def predict(self, x: pd.DataFrame) -> list[str]: ...


@dataclass(frozen=True)
class Score:
    train: float
    test: float


def evaluate(model: Classifier, split: TrainTestSplit) -> Score:
    model.fit(split.x_train, split.t_train)
    return Score(
        train=accuracy(model.predict(split.x_train), split.t_train.to_list()),
        test=accuracy(model.predict(split.x_test), split.t_test.to_list()),
    )

File: apps/node/src/chapter10/classifier.ts
import { accuracy } from "../chapter01/kinoko-takenoko.ts";
import type { TrainTestSplit } from "../chapter02/iris-preprocessing.ts";

/** fit で学習し、predict でラベルを予測する分類モデル */
export interface Classifier<K extends string> {
  fit(x: readonly Record<K, number>[], t: readonly string[]): this;
  predict(x: readonly Record<K, number>[]): string[];
}

export interface Score {
  train: number;
  test: number;
}

export function evaluate<K extends string>(
  model: Classifier<K>,
  split: TrainTestSplit<Record<K, number>, string>,
): Score {
  model.fit(split.xTrain, split.tTrain);
  return {
    train: accuracy(model.predict(split.xTrain), split.tTrain),
    test: accuracy(model.predict(split.xTest), split.tTest),
  };
}
